ProcessList.c

```diff
@@ -112,16 +115,73 @@ int ProcessList_size(ProcessList* this) {
    return (Vector_size(this->processes));
 }
 
-static void ProcessList_buildTree(ProcessList* this, pid_t pid, int level, int indent, int direction, bool show) {
+static int ProcessList_updateTreeSetLayer(ProcessList* this, unsigned int deep, unsigned int lft, unsigned int rht, int* index, int* tree_index, int indent) {
+   Vector* layer = Vector_new(this->processes->type, false, DEFAULT_SIZE);
+   for (int i = 0; i < Vector_size(this->processes); i++) {
+      Process* proc = (Process*)Vector_get(this->processes, i);
+      if (proc->tree_deep == deep && proc->tree_lft > lft && proc->tree_rht < rht) {
+         Vector_add(layer, proc);
+      }
+   }
```

Comment:
IF you ensure the nested set is up-to-date you can optimize this to start from `parent->tree_left+1` and skipping to `currProcess->tree_right+1` until you reach `parent->tree_right`. Also initial size for that Vector could be estimated (reserved) as `(parent->tree_right - parent->tree_left)/2` and you would never need to re-allocate while extracting that list.

Reply:
An optimisation could be possible if we would know that `this->processes` was sorted already by `TREE_SET`. Since currently I assume it's may be false

Reply:
Yes. But only given the dirent child processes for one parent. Retrieving all processes on level x still would have you search recursively everything on layers above. But this should be linear complexity.

Reply:
Look at the comment.